linked_list/linked_list.py

```diff
@@ -1,75 +1,181 @@
 
 # Defines a node in the singly linked list
+# from ast import Pass
+
+
 class Node:
 
     def __init__(self, value, next_node = None):
         self.value = value
         self.next = next_node
+        # self.previous = previous_node
 
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None 
+        # keep the head private. Not accessible outside this class
+        # self.length = 0 # keep the length private. Not accessible outside this class
+        # if self.head == None:
+        #     self.head = Node(None)
+        #     self.length = 0
+        # else:
+        #     self.length = 1
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
 
+        if self.head == None:
+            return None
+        else: 
+            return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+
+        if self.head == None:
+            self.head = Node(value)
+        else:
+            new_node = Node(value)
+            new_node.next = self.head
+            self.head = new_node
+            return self.head
+
+        # if self.head == None:
+        #     return None 
+
+        # self.head = Node(value, self.head)
+        # self.length += 1
+        # return self.head
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)

```

Comment:
⏱ Time complexity is O(n) since in the worst case the node with the value you are searching for will be the last node in the list, and you will have to iterate through every node in the list.